docs/proposals/server-side-apply.md

```diff
@@ -0,0 +1,194 @@
+---
+title: Server-Side Apply
+
+authors:
+- "@leoluz"
+
+sponsors:
+- TBD
+
+reviewers:
+- TBD
+
+approvers:
+- TBD
+
+creation-date: 2022-03-17
+
+---
+
+# Server-Side Apply support for ArgoCD
+
+[Server-Side Apply (SSA)][1] allows calculating the final patch to update
+resources in Kubernetes in the server instead of the client. This proposal
+describes how ArgoCD can leverage SSA during syncs.
+
+* [Open Questions](#open-questions)
+    * [[Q-1] How to handle conflicts?](#q-1-how-to-handle-conflicts)
+    * [[Q-2] Should we support multiple managers?](#q-2-should-we-support-multiple-managers)
+* [ Summary ](#summary)
+* [Motivation](#motivation)
+    * [Better interoperability with Admission Controllers](#better-interoperability-with-admission-controllers)
+    * [Better resource conflict management](#better-resource-conflict-management)
+    * [Better CRD support](#better-crd-support)
+* [Goals](#goals)
+* [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Non-Functional Requirements](#non-functional-requirements)
+    * [Use cases](#use-cases)
+        * [[UC-1]: enable SSA at the controller level](#uc-1-as-a-user-i-would-like-enable-ssa-at-the-controller-level-so-all-application-are-applied-server-side)
+        * [[UC-2]: enable SSA at the Application level](#uc-2-as-a-user-i-would-like-enable-ssa-at-the-application-level-so-all-resources-are-applied-server-side)
+        * [[UC-3]: enable SSA at the resource level](#uc-3-as-a-user-i-would-like-enable-ssa-at-the-resource-level-so-only-a-single-manifest-is-applied-server-side)
+    * [Security Considerations](#security-considerations)
+    * [Risks and Mitigations](#risks-and-mitigations)
+    * [Upgrade / Downgrade](#upgrade--downgrade)
+* [Drawbacks](#drawbacks)
+
+---
+
+## Open Questions
+
+### [Q-1] How to handle conflicts?
+When SSA is enabled, the server may return field conflicts with other managers.
+What ArgoCD controller should do in case of conflict? Just force the sync and
+log warnings (like some other controllers do?)
+
+### [Q-2] Should we support multiple managers?
+Should Server-Side Apply support in ArgoCD be implemented allowing multiple
+managers for the same controller? ([more details][10])
+
+## Summary
+
+ArgoCD can benefit from [Server-Side Apply][1] during syncs. A few
+improvements to consider:
+
+- More reliable dry-runs (as admission controller is executed) ([ISSUE-804][5])
+- [Syncs always run][2] mutating webhooks (even without diff)
+- [Fix big CRD][3] sync issues
```

Comment:
I think kubectl still going to create `last-applied-configuration` annotation. So this issue won't be solved?

Reply:
From the SSA Kubernetes proposal ([KEP-555](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/555-server-side-apply/README.md#upgrade--downgrade-strategy)):

> To upgrade the last-applied-configuration annotation, the following procedure will be used:
> 1. Identify if the server-side apply is from the kubectl user-agent
> 2. Identify if the server-side apply would result in a conflict
> 3. Create a fieldset from the last-applied-configuration annotation.
> 4. Remove all fields from the last-applied-configuration annotation that are added, missing, or different than the corresponding field of the live object. Because the fields have changed, client-side apply does not own them.
> 5. Compare the "last-applied" fieldset to the conflict fieldset. Take the difference as the new conflict fieldset. If the conflict fieldset is empty, then the conflicts are allowed and we force the server-side apply. If the conflict fieldset is not empty, then return the conflict fieldset.

My understanding is that in step 4 when a resource is applied server-side the fields ownership is transferred from client (kubectl) to server. Kubectl uses the `last-applied-configuration` to identify field ownership and check if state changed. This means that the resulting `last-applied-configuration` will be always smaller than the original value as the ownership was transferred.